maskit/masks.py

```diff
@@ -0,0 +1,242 @@
+        return clone
+
+
+class MaskedCircuit(object):
+    """
+    A MaskedCircuit supports masking of different components including wires,
+    layers, and parameters.
+    """
+
+    __slots__ = (
+        "_layer_mask",
+        "_wire_mask",
+        "_parameter_mask",
+        "parameters",
+    )
+
+    def __init__(self, parameters: np.ndarray, layers: int, wires: int):
+        assert (
+            layers == parameters.shape[0]
+        ), "First dimension of parameters shape must be equal to number of layers"
+        assert (
+            wires == parameters.shape[1]
+        ), "Second dimension of parameters shape must be equal to number of wires"
+        self.parameters = parameters
+        self._parameter_mask = Mask(shape=parameters.shape)
+        self._layer_mask = Mask(shape=(layers,))
+        self._wire_mask = Mask(shape=(wires,))
+
+    @property
+    def mask(self) -> np.ndarray:
+        """
+        Accumulated mask of layer, wire, and parameter masks.
+        Note that this mask is readonly.
+        """
+        mask = self.parameter_mask.copy()
+        mask[self.layer_mask] = True
+        mask[:, self.wire_mask] = True
+        return mask
+
+    @property
+    def layer_mask(self):
+        """Returns the encapsulated layer mask."""
+        return self._layer_mask.mask
+
+    @property
+    def wire_mask(self):
+        """Returns the encapsulated wire mask."""
+        return self._wire_mask.mask
+
+    @property
+    def parameter_mask(self):
+        """Returns the encapsulated parameter mask."""
+        return self._parameter_mask.mask
+
+    def perturb(
+        self,
+        axis: PerturbationAxis,
+        amount: Optional[int] = None,
+        mode: PerturbationMode = PerturbationMode.INVERT,
+    ):
+        """
+        Perturbs the MaskedCircuit for a given ``axis`` that is of type
+        :py:class:`~.PerturbationAxis`. The perturbation is applied ``amount``times
+        and depends on the given ``mode`` of type :py:class:`~.PerturbationMode`.
+        If no amount is given, that is ``amount=None``, a random ``amount`` is
+        determined given by the actual size of the py:attr:`~.mask`. The ``amount``
+        is automatically limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to perturb, defaults to None
+        :param axis: Which mask to perturb
+        :param mode: How to perturb, defaults to PerturbationMode.INVERT
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert mode in list(PerturbationMode), "The selected mode is not supported."
+        if amount == 0:
+            return
+        if axis == PerturbationAxis.LAYERS:
+            mask_fn = self._mask_layers
+        elif axis == PerturbationAxis.WIRES:
+            mask_fn = self._mask_wires
+        elif axis == PerturbationAxis.RANDOM:  # Axis is on parameters
+            mask_fn = self._mask_parameters
```

Comment:
This just tripped me up since I read them as the fields – e.g. ``_mask_layers`` as ``_layer_mask`` – before realising they are actually methods. You could remove the intermediate methods and directly fetch the ``perturb`` method here.
```suggestion
        if axis == PerturbationAxis.LAYERS:
            mask_fn = self._layer_mask.perturb
        elif axis == PerturbationAxis.WIRES:
            mask_fn = self._wire_mask.perturb
        elif axis == PerturbationAxis.RANDOM:  # Axis is on all parameters
            mask_fn = self._parameter_mask.perturb
```
As an alternative, select the mask (``mask = self._layer_mask``) and then call perturb on the result (``mask.perturb(amount=amount, mode=mode)``).

Reply:
Good idea, thanks!